### PR TITLE
Improve checkstyle performance and readability

### LINKED ISSUE #53921: Fix Checkstyle SnippetLength Suppressions in HLRC ITs
We have this checkstyle snippet length check in place but it (it's multi-line regex) produces stack overflows here and there so we excluded it for a large number number of these ITs rendering the check somewhat pointless.

see excludes at: https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/resources/checkstyle_suppressions.xml#L28

We should fix the regex to not overflow and remove these exclusions.

Relates https://github.com/elastic/elasticsearch/pull/53840

### PR DESCRIPTION
**Pull request by nik9000**
Drop a nasty regex in our checkstyle config that I wrote a long time ago
in favor of a checkstyle extension. This is better because:
* It is faster. It saves a little more than a minute across the entire
  build.
* It is easier to read. Who knew 100 lines of Java would be easier to
  read than a regex, but it is.
* It has tests.

Closes #53921

### CODE REVIEW

File: checkstyle/licenses/checkstyle-LICENSE.txt [hunk 1]
```diff
@@ -0,0 +1,503 @@
+                  GNU LESSER GENERAL PUBLIC LICENSE
```
Comment (nik9000) [inline checkstyle/licenses/checkstyle-LICENSE.txt]: @tomcallahan could you take a look at this? We depend on checkstyle in the build and this extension is part of the build.

File: settings.gradle [hunk 2]
```diff
@@ -8,6 +8,7 @@ rootProject.name = dirName
 List projects = [
   'build-tools',
   'build-tools:reaper',
+  'checkstyle',
```
Comment (nik9000) [inline settings.gradle]: I did not put this in buildsrc because it *looks* like the logger usage check to me. I can certainly move it if you'd feel more comfortable with it in the buildSrc but I might need tips on how to reference it in that case.

**Comment by rjernst** [comment]
Please do put the project under buildSrc. We should arrive to have anything build related under there, even in subprojects like reaper.

File: settings.gradle (hunk 2 above)
Comment (mark-vieira) [reply settings.gradle]: It probably should move as well. It's not production code, but the current configuration is probably just historical.

File: checkstyle/src/main/java/org/elasticsearch/checkstyle/SnippetLengthCheck.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        while (lines.hasNext()) {
+            String line = lines.next();
+            if (end.matcher(line).matches()) {
+                return;
+            }
+            if (line.isEmpty()) {
+                continue;
+            }
+            if (false == line.startsWith(leadingSpaces)) {
+                log.accept(lines.lastLineNumber, "snippet line should start with [" + leadingSpaces + "]");
```
Comment (mark-vieira) [inline checkstyle/src/main/java/org/elasticsearch/checkstyle/SnippetLengthCheck.java]: This log message might be a little confusing as we are just going to print n spaces. Perhaps we should reword this to say that "snippet line should start with n leading spaces" to make it clear that we are talking indent width.

File: checkstyle/src/main/java/org/elasticsearch/checkstyle/SnippetLengthCheck.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck;
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+import com.puppycrawl.tools.checkstyle.api.FileText;
+
+/**
+ * Checks the snippets included in the docs aren't too wide to fit on
+ * the page.
+ */
+public class SnippetLengthCheck extends AbstractFileSetCheck {
+    private static final Pattern START = Pattern.compile("^( *)//\\s*tag::(.+?)\\s*$", Pattern.MULTILINE);
```
Comment (mark-vieira) [inline checkstyle/src/main/java/org/elasticsearch/checkstyle/SnippetLengthCheck.java]: We are we using multiline matchers when we are always matching on individual lines?

File: checkstyle/src/main/java/org/elasticsearch/checkstyle/SnippetLengthCheck.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+     * The maximum width that a snippet may have.
+     */
+    public void setMax(int max) {
+        this.max = max;
+    }
+
+    @Override
+    @SuppressForbidden(reason = "Checkstyle's API contain File")
+    protected void processFiltered(File file, FileText fileText) throws CheckstyleException {
+        checkFile((line, message) -> log(line, message), max, fileText.toLinesArray());
```
Comment (mark-vieira) [inline checkstyle/src/main/java/org/elasticsearch/checkstyle/SnippetLengthCheck.java]: Seems a bit unnecessarily to plum this log function which just maps 1:1 with the `log()` method of the parent class. Is this just so that these methods can be `static`? Is there any real benefit to that? Does checkstyle create multiple instances of this?

File: checkstyle/src/main/java/org/elasticsearch/checkstyle/SnippetLengthCheck.java (hunk 4 above)
Comment (nik9000) [reply checkstyle/src/main/java/org/elasticsearch/checkstyle/SnippetLengthCheck.java]: That makes sense!

File: checkstyle/src/main/java/org/elasticsearch/checkstyle/SnippetLengthCheck.java (hunk 5 above)
Comment (nik9000) [reply checkstyle/src/main/java/org/elasticsearch/checkstyle/SnippetLengthCheck.java]: It means I don't have to figure out anything about the way checkstyle works to write the test. I could probably rework it so the test is more checkstyle-native if you'd prefer.

**Comment by nik9000** [comment]
Update for those playing along at home: @mark-vieira is going to see if he can find a nice way to move the new project into `buildSrc`.

**Comment by mark-vieira** [comment]
This PR is superseded by #54308.
